# **Tarea 02**

# Pandas 01 **Trabajando con Pandas DataFrames en Python**

In [2]:
import pandas as pd

In [3]:
data=pd.read_csv("pandas_data/pandas_data/surveys.csv")

> ## 01) DataFrames
>
> Usando el archvivo **surveys.csv** y definiendo el data frame como `surveys_df`, ejecuta los atributos y métodos siguientes
> y observa que regresan.
>
> 1. `surveys_df.columns`
> 2. `surveys_df.shape` Toma nota de la salida de`shape` - ¿Qué formato tiene la salida
>    del atributo que regresa la forma de un DataFrame?
> 3. `surveys_df.head()` También ejecuta `surveys_df.head(15)` ¿qué hace esto?
> 4. `surveys_df.tail()`

In [4]:
data.shape

(35549, 9)

1. el código me muestra todos los nombres de las columnas que contiene el dataframe
2. me muestra la cantidad de filas y columnas del dataframe (filas, columnas)
3. cuando ejecuto .head() me muestra las primeras 5 filas del dataframe, mientras que con .head(15) me muestra las primeras 15 filas
4. me muestra las últimas 5 filas del dataframe

> ## 02) Descripción de datos
>
> 1. ¿Cuántos individuos son hembras `F` y cuántos son machos `M`?
> 2. Qué pasa cuando agrupas sobre dos columnas usando el siguiente enunciado
>    y después tomas los valores medios:
>   - `grouped_data2 = surveys_df.groupby(['plot_id','sex'])`
>   - `grouped_data2.mean()`
> 3. Calcula las estadísticas descriptivas del peso (weight) para cada sitio (plot_id). Sugerencia:
>    puedes usar la siguiente sintaxis para solo crear las estadísticas para una
>   columna de tus datos
>   `by_site['weight'].describe()`

In [8]:
data["sex"].value_counts()

sex
M    17348
F    15690
Name: count, dtype: int64

In [11]:
data_ = data.groupby("sex")["sex"].count()
data_

sex
F    15690
M    17348
Name: sex, dtype: int64

In [12]:
data_unida = data.groupby(['plot_id','sex'])
data_unida.mean()

TypeError: agg function failed [how->mean,dtype->object]

In [7]:
data.groupby("plot_id")["weight"].describe()

count       mean        std  min   25%   50%   75%    max
plot_id                                                            
1        1903.0  51.822911  38.176670  4.0  30.0  44.0  53.0  231.0
2        2074.0  52.251688  46.503602  5.0  24.0  41.0  50.0  278.0
3        1710.0  32.654386  35.641630  4.0  14.0  23.0  36.0  250.0
4        1866.0  47.928189  32.886598  4.0  30.0  43.0  50.0  200.0
5        1092.0  40.947802  34.086616  5.0  21.0  37.0  48.0  248.0
6        1463.0  36.738893  30.648310  5.0  18.0  30.0  45.0  243.0
7         638.0  20.663009  21.315325  4.0  11.0  17.0  23.0  235.0
8        1781.0  47.758001  33.192194  5.0  26.0  44.0  51.0  178.0
9        1811.0  51.432358  33.724726  6.0  36.0  45.0  50.0  275.0
10        279.0  18.541219  20.290806  4.0  10.0  12.0  21.0  237.0
11       1793.0  43.451757  28.975514  5.0  26.0  42.0  48.0  212.0
12       2219.0  49.496169  41.630035  6.0  26.0  42.0  50.0  280.0
13       1371.0  40.445660  34.042767  5.0  20.5  33.0  45.0  241.0
14       1728.0  46.277199  27.570389  5.0  36.0  44.0  49.0  222.0
15        869.0  27.042578  35.178142  4.0  11.0  18.0  26.0  259.0
16        480.0  24.585417  17.682334  4.0  12.0  20.0  34.0  158.0
17       1893.0  47.889593  35.802399  4.0  27.0  42.0  50.0  216.0
18       1351.0  40.005922  38.480856  5.0  17.5  30.0  44.0  256.0
19       1084.0  21.105166  13.269840  4.0  11.0  19.0  27.0  139.0
20       1222.0  48.665303  50.111539  5.0  17.0  31.0  47.0  223.0
21       1029.0  24.627794  21.199819  4.0  10.0  22.0  31.0  190.0
22       1298.0  54.146379  38.743967  5.0  29.0  42.0  54.0  212.0
23        369.0  19.634146  18.382678  4.0  10.0  14.0  23.0  199.0
24        960.0  43.679167  45.936588  4.0  19.0  27.5  45.0  251.0

1. Hay 15690 hembras y 17348 machos
2. a

> ## 03) Haciendo una lista
>
> ¿Qué otra manera hay de crear una lista de especies (species_id) y asociarle `count` de las
>  muestras  de los datos? Sugerencia: puedes hacer ejecutar las funciones
>  `count`, `min`, etc. en un **DataFrame** groupby de la misma manera de la que
>  se hace en un DataFrame

In [1]:
data_ = data.groupby("species_id").count()
data_

NameError: name 'data' is not defined

# Pandas 02 **Indexación, segmentación y creación de subconjuntos a partir de DataFrames en Python**

> ## Rangos
>
> 1. ¿Qué ocurre al ejecutar el siguiente código?:
>
>    - `surveys_df[0:1]`
>    - `surveys_df[:4]`
>    - `surveys_df[:-1]`
>
> 2. ¿Qué pasa al ejecutar este otro?:
>
>    - `dat.iloc[0:4, 1:4]`
>    - `dat.loc[0:4, 1:4]`
>
> - ¿Qué diferencia observas entre los resultados de los comandos inmediatamente anteriores?

> ## Consultas
>
> 1. Selecciona un subconjunto de filas, en el `DataFrame` `surveys_df`, que contenga datos desde
>   el año 1999 y que contenga valores en `weight` menores o iguales a 8. ¿Cuántas
>   filas obtuviste como resultado? ¿Cuántas filas obtuvo tu compañera?
>
> 2. Puedes usar la función `isin` de Python para hacer una consulta a un `DataFrame`, basada en
>   una lista de valores según se muestra a continuación:
>
>    
>    `surveys_df[surveys_df['species_id'].isin([listGoesHere])]`
>    

>
>   Usa la función `isin` para encontrar todos los `plots` que contienen una especie en particular
>   en el `DataFrame` `surveys_df`. ¿Cuántos registros contienen esos valores?
>
> 3. Experimenta con otras consultas. Crea una consulta que encuentre todas las filas con el
>   valor de `weight` mayor o igual a 0.
>
> 4. El símbolo `!=` puede ser usado en Python para obtener lo **opuesto** a los
>   datos seleccionados que hayas especificado. Es equivalente a **no esta en**.
>   Escribe una consulta que seleccione todas las filas con `sex` diferente de 'M' o 'F' en
>   los datos de `surveys_df`.

> ##  Revisando todo lo aprendido
>
> 1. Crea un nuevo objeto `DataFrame` que solamente contenga observaciones cuyos valores en la columna `sex`
>   **no** sean `female` o `male`. Asigna cada valor de `sex` en el nuevo `DataFrame` a
>   un nuevo valor de 'x'. Determina el número total de valores `null` en el subconjunto.
>
> 2. Crea un nuevo objeto `DataFrame` que contenga solo observaciones cuyos valores en la columna `sex` sean `male`
>   o `female` y en los cuales el valor de `weight` sea mayor que 0. 

# Pandas 3 **Tipos de datos y formatos**

> # Contando
>
> Cuenta el número de valores perdidos por columna. Sugerencia: el método .count() te proporciona el número de observaciones que no son NA por columna. Examina el método .isnull().

> # Cambiando tipos
>
> Intenta convertir la columna plot_id a float usando
>
> ` surveys_df.plot_id.astype("float")`
> 
> A continuación, intenta convertir weight (peso) en un integer. ¿Qué te dice Pandas? ¿Qué es lo que va mal ahí?

 # Pandas 04 **Combinando DataFrames con Pandas**

> ## Distribuciones
>
> Cree un nuevo DataFrame uniendo los contenidos de survey.csv y species.csv. Luego calcula:
>
>   - taxa por parcela
>   - taxa por sexo por parcela


> ## Índice de Diversidad
> - En la carpeta de datos, hay un archivo plots.csv que contiene información sobre el tipo asociado con cada parcela. Usa esos datos para resumir el número de parcelas por tipo de parcela.
>  - Calcula un índice de diversidad de su elección para control vs parcelas de rodamiento de roedores. El índice debe considerar tanto la abundancia de especies como el número de especies. Puedes optar por utilizar el simple índice de biodiversidad descrito aquí que calcula la diversidad como:
> el número de especies en la parcela / el número total de individuos en la parcela = índice de biodiversidad.